Limit CPU features through a hint

madebr opened the pull request:
- Introduces a `SDL_HINT_CPU_FEATURE_MASK` hint with which you can limit supported cpu features.
- Add another run of testautomation and testplatform to the test matrix, but without simd support.
- Always build scalar audio converters, since it is now possible to e.g. not have sse2 on x86_64.
- Reset cpuinfo flags on `SDL_Quit`

Things to consider when reviewing:
- is the hint correctly named?
- for all CPU features, you currently need to run with `SDL_CPU_FEATURES=sse,sse2,sse3,sse41,sse42`. Should we assume when you pass `sse42`, you also want sse1 to sse4.1?
- please review the hint documentation
- ~The hint is only parsed once. Should it be used through `SDL_AddHintCallback` instead? This would be useful for GUI options, or for doing performance testing with one non-forking executable.~ Not needed anymore by resetting cpuinfo in `SDL_Quit`.

## Existing Issue(s)
https://github.com/libsdl-org/SDL/issues/8994

madebr commented on include/SDL3/SDL_hints.h:
```diff
@@ -371,6 +371,35 @@ extern "C" {
+ *
+ *  This hint is available since SDL 3.0.0.
+ */
+#define SDL_HINT_CPU_FEATURE_MASK "SDL_CPU_FEATURE_MASK"
```
`SDL_CPU_FEATURES_MASK` or `SDL_CPU_FEATURE_MASK`?

slouken replied:
`SDL_CPU_FEATURE_MASK`

slouken commented:
I wonder if we should invert this, and it should be a hint of CPU features to disable, with "all" disabling all CPU features? That would make more sense in the context that you're testing what's currently available on the CPU, and resolve the question of whether earlier features should be enabled when you enable a later instruction set.

madebr commented:
> I wonder if we should invert this, and it should be a hint of CPU features to disable, with "all" disabling all CPU features? That would make more sense in the context that you're testing what's currently available on the CPU, and resolve the question of whether earlier features should be enabled when you enable a later instruction set.

This would make this the first "negative" hint.

Perhaps, we can (additionally) allow the following syntax?
- `"all,-sse2"` -> mask = 0xffffffff & ~SSE2
- `"none,+sse2,+sse3"` -> mask = SSE2 | SSE3

slouken commented:
Sure, that sounds fine to me.

madebr commented:
The latest revision allows this usage
```
SDL_CPU_FEATURE_MASK=all,-sse2,-sse41,-avx test/testplatform
...
INFO: AltiVec not detected
INFO: MMX detected
INFO: SSE detected
INFO: SSE2 not detected
INFO: SSE3 detected
INFO: SSE4.1 not detected
INFO: SSE4.2 detected
INFO: AVX not detected
INFO: AVX2 detected
INFO: AVX-512F not detected
INFO: ARM SIMD not detected
INFO: NEON not detected
INFO: LSX not detected
INFO: LASX not detected
...
```

madebr commented:
I converted to a draft because setting `SDL_CPU_FEATURE_MASK` to an empty value does not work on Windows: I expect all features to be disabled, but they are enabled.

madebr commented:
> I converted to a draft because setting `SDL_CPU_FEATURE_MASK` to an empty value does not work on Windows: I expect all features to be disabled, but they are enabled.

Looks like on Windows, setting the [`ENVIRONMENT` property](https://cmake.org/cmake/help/latest/prop_test/ENVIRONMENT.html)) on a test to an empty value removes the environment variable, thus resetting it.

So I changed the behavior of the hint to the following:
- when the hit is not set, or empty, all options are enabled
- you can subtract options by prepending with `'-'`
- add options by prepending with `'+'`

Some examples:
- enable all features: `(not set)`, `""`, `"all"` or `"+all"`
- disable all features: `"-all"`
- disable only sse2: `"-sse2"`
- enable only sse2: `"-all,+sse2"`


> I wonder if we should invert this, and it should be a hint of CPU features to disable, with "all" disabling all CPU features? That would make more sense in the context that you're testing what's currently available on the CPU, and resolve the question of whether earlier features should be enabled when you enable a later instruction set.

This is now the new behavior.
